compiler_opt/es/blackbox_learner.py

```diff
@@ -0,0 +1,335 @@
+                                 rewards: List[Optional[float]]):
+  """Remove perturbations that were skipped during the training step.
+
+  Perturbations may be skipped due to an early exit condition or a server error
+  (clang timeout, malformed training example, etc). The blackbox optimizer
+  assumes that each perturbations has a valid reward, so we must remove any of
+  these "skipped" perturbations.
+
+  Pruning occurs in-place.
+
+  Args:
+    perturbations: the model perturbations used for the ES training step.
+    rewards: the rewards for each perturbation.
+
+  Returns:
+    The number of perturbations that were pruned.
+  """
+  indices_to_prune = []
+  for i, reward in enumerate(rewards):
+    if reward is None:
+      indices_to_prune.append(i)
+
+  # Iterate in reverse so that the indices remain valid
+  for i in reversed(indices_to_prune):
+    del perturbations[i]
+    del rewards[i]
+
+  return len(indices_to_prune)
+
+
+class PolicySaverCallableType(Protocol):
+  """Protocol for the policy saver function.
+  A Protocol is required to type annotate
+  the function with keyword arguments"""
+
+  def __call__(self, parameters: npt.NDArray[np.float32],
+               policy_name: str) -> None:
+    ...
+
+
+class BlackboxLearner:
+  """Implementation of blackbox learning."""
+
+  def __init__(self,
+               blackbox_opt: blackbox_optimizers.BlackboxOptimizer,
+               sampler: corpus.Corpus,
+               tf_policy_path: str,
+               output_dir: str,
+               policy_saver_fn: PolicySaverCallableType,
+               model_weights: npt.NDArray[np.float32],
+               config: BlackboxLearnerConfig,
+               initial_step: int = 0,
+               deadline: float = 30.0):
+    """Construct a BlackboxLeaner.
+
+    Args:
+      blackbox_opt: the blackbox optimizer to use
+      train_sampler: corpus_sampler for training data.
+      tf_policy_path: where to write the tf policy
+      output_dir: the directory to write all outputs
+      policy_saver_fn: function to save a policy to cns
+      model_weights: the weights of the current model
+      config: configuration for blackbox optimization.
+      stubs: grpc stubs to inlining/regalloc servers
+      initial_step: the initial step for learning.
+      deadline: the deadline in seconds for requests to the inlining server.
+    """
+    self._blackbox_opt = blackbox_opt
+    self._sampler = sampler
+    self._tf_policy_path = tf_policy_path
+    self._output_dir = output_dir
+    self._policy_saver_fn = policy_saver_fn
+    self._model_weights = model_weights
+    self._config = config
+    self._step = initial_step
+    self._deadline = deadline
+
+    # While we're waiting for the ES requests, we can
+    # collect samples for the next round of training.
+    self._samples = []
+
+    self._summary_writer = tf.summary.create_file_writer(output_dir)
+
+  def _get_perturbations(self, seed=None) -> List[npt.NDArray[np.float32]]:
```

Comment:
you could make `seed` a member of `BlackboxLearner` and then testing `run_step` would also be deterministic.